Modify title and favicon when item in progress (issue #2392)

jungleBadger opened the pull request:
## Summary

### Description

Two methods (`updatePageTitle` and `updatePageFavicon`) were created and executed within the WebSocket handler.

When a Queue item's status changes, the methods append a `(1) ` prefix and modify the favicon, respectively. I included an `id` attribute to the `link` element to ensure the query wouldn't fail or collide.

After Hipsterusername's suggestion, I modified the `updatePageTitle` method to display the active queue length instead of (1). 

### Files changed:
1. `invokeai/frontend/web/src/app/store/middleware/listenerMiddleware/listeners/socketio/socketQueueItemStatusChanged.ts`
2. `invokeai/frontend/web/index.html`

### Files added
1. `invokeai/frontend/web/public/assets/images/invoke-alert-favicon.svg`

### Demo

https://github.com/invoke-ai/InvokeAI/assets/12788435/d60db7f9-0e7e-4204-9914-ab968e03f165

(I will update the final demo video with all suggestions implemented once it's done)

### Open questions

1. Do you have any recommendations on how to create automated tests?
3. Should I split the methods into a dedicated file and import them on the Socket handler? I thought about doing so, but it would make more sense to go with the simplest approach and adjust if needed.

## Related Issues / Discussions

Address issue #2392 

## QA Instructions

1. Ensure there is an active model
4. Input a prompt
5. Click on the `invoke` button
6. Observe the title and favicon

## Merge Plan

-

## Checklist

- [X] _The PR has a short but descriptive title, suitable for a changelog_
- [ ] _Tests added / updated (if applicable)_
- [ ] _Documentation added / updated (if applicable)_

hipsterusername commented:
Neat! And even some banana sushi!?! 👏 

Is there a reason for (1) instead of the current active queue?

jungleBadger commented:
> Neat! And even some banana sushi!?! 👏
> 
> Is there a reason for (1) instead of the current active queue?

I couldn't think of a better combination; as a good Brazilian, I love bananas and sushis!

The active queue number is always increasing, so I would need to implement an event to clean up the icon and favicon once the user returns to the application, similar to what happens on chat applications. 

It would be simpler to keep track of running items and clear them upon completion. Still, the current implementation can be flawed if parallel operations are to be supported.

I'm open if you have any suggestions to improve this feature.

jungleBadger commented:
@hipsterusername Thanks for your suggestion; I reviewed the code and implemented a solution similar to what I found in the `QueueActionsMenuButton.tsx` script to determine the `activeQueueLength`

### Demo

https://github.com/invoke-ai/InvokeAI/assets/12788435/502da3a9-921d-4581-9254-426780f1855b

### Questions

1. I implemented a small regular expression to clean up the title. I did so to avoid hardcoding the page title in the code or implementing some separator that could cause problems later on. However, I could split the string in the `) ` pattern and go from there. Please let me know if RegEx is not desirable, and I can try to adapt.
2. Do you have any tips on modularizing and creating unit testing considering the project stack?

psychedelicious commented on invokeai/frontend/web/public/assets/images/invoke-alert-favicon.svg:
![invoke-alert-favicon](https://github.com/invoke-ai/InvokeAI/assets/4822129/536f4afa-4a53-41ea-8134-d00599cc2ad9)

Here's a slightly modified version that includes a tiny bit of padding around the red dot and uses our brand red color.

jungleBadger replied:
Thanks for the guidance. Based on the suggestions, I have:

1. Removed the previous implementation from the `socketQueueItemStatusChanged` script
2. Created a new file named `useSyncQueueStatus` within `invokeai/frontend/web/src/app/hooks`
3. Imported the file in App.ts
4. Modified the alert SVG by including the black border and correct red color

psychedelicious reviewed (changes requested):
Cool!

The implementation isn't quite right. Currently, we are imperatively updating the title and favicon in the socket handler, but the queue status can be changed by other code paths. For example, if you queue 5 items, then clear the queue, the count and icon do not reset as expected, because that query handler isn't updating the title and favicon.

The cache for `useGetQueueStatusQuery` can be considered the source of truth for the queue status. The socket events and endpoints all sync to the cache. For example, in the listener you updated, you can see we update the query cache.

I'd suggest a hook that uses `useEffect` to sync the title and favicon, using the data from that query cache. 

- The hook would be a singleton. It's fine to just add a docstring to the hook saying to consider it a singleton, instead of programmatically enforcing it.
- Define constants as module-scoped variables outside the hook (`document.title` can be referenced for the initial value outside the hook)
- Use a type guard that checks if `faviconEl` is an instance of `HTMLLinkElement` instead of casting it
- Call the hook once in `App.tsx`

psychedelicious commented:
We don't have an e2e test setup for the frontend. Don't worry about tests.

psychedelicious reviewed (approved):
Looks great, thank you!
